iroha/src/sumeragi.rs

```diff
@@ -164,82 +164,130 @@ impl Sumeragi {
-                            if *peer != peer_id {
-                                send_futures.push(
-                                    Message::NoTransactionReceiptReceived(no_tx_receipt.clone())
-                                        .send_to(peer),
-                                );
-                            }
+            }
+            let transactions_awaiting_receipts = self.transactions_awaiting_receipts.clone();
+            let mut no_tx_receipt = NoTransactionReceiptReceived::new(
+                &transaction,
+                self.network_topology.leader().clone(),
+                self.latest_block_hash,
+                self.number_of_view_changes,
+            );
+            let role = self.network_topology.role(&self.peer_id);
+            if role == Role::ValidatingPeer || role == Role::ProxyTail {
+                no_tx_receipt = no_tx_receipt
+                    .sign(&self.key_pair)
+                    .expect("Failed to put first signature.");
+            }
+            let recipient_peers = self.network_topology.sorted_peers.clone();
+            let transaction_hash = transaction.hash();
+            let peer_id = self.peer_id.clone();
+            let tx_receipt_time = self.tx_receipt_time;
+            let _ = async_std::task::spawn(async move {
+                async_std::task::sleep(tx_receipt_time).await;
+                if transactions_awaiting_receipts
+                    .write()
+                    .await
+                    .contains(&transaction_hash)
+                {
+                    let mut send_futures = Vec::new();
+                    for peer in &recipient_peers {
+                        if *peer != peer_id {
+                            send_futures.push(
+                                Message::NoTransactionReceiptReceived(no_tx_receipt.clone())
+                                    .send_to(peer),
+                            );
                         }
-                        let results = futures::future::join_all(send_futures).await;
-                        results
-                            .iter()
-                            .filter(|result| result.is_err())
-                            .for_each(|error_result| {
-                                log::error!(
-                                    "Failed to send NoTransactionReceiptReceived message to peers: {:?}",
-                                    error_result
-                                )
-                            });
                     }
-                });
+                    let results = futures::future::join_all(send_futures).await;
+                    results
+                        .iter()
+                        .filter(|result| result.is_err())
+                        .for_each(|error_result| {
+                            log::error!(
+                                "Failed to send NoTransactionReceiptReceived message to peers: {:?}",
+                                error_result
```

Comment:
Would it be better to unwrap result here?

Reply:
Or i guess `filter` can be substituted with `filter_map` with `.err` inside closure.

Reply:
Here it is part of the consensus. In general we would not want to unwrap or fail with error here. The network and peers are unreliable, so therefore we simply inform if some of them behave strangely instead of stopping the peer.

Reply:
It is the task of the consensus to let the blockchain function even if some of the peers are faulty.

Reply:
We would never fail in unwraping here as we filtered only errors here. (Sorry I meant `.unwrap_err`).

You have some similar code bellow:
``` rust
results
    .into_iter()
    .filter_map(|result| result.err())
    .for_each(|error| {
        log::error!(
            "Failed to send BlockCreated messages from {}: {:?}",
             this_peer.address,
             error
        )
    });
```
In for each there are only errors now so we will just print it out.

Reply:
Ah, I see, yes, seems like I misunderstood you. Okay, thanks, will correct it here.